satisfy/src/ConCat/BuildDictionary.hs

```diff
@@ -180,54 +187,49 @@ buildDictionary' env dflags guts evIds evar =
 
 -- TODO: Try to combine the two runTcM calls.
 
-buildDictionary :: HscEnv -> DynFlags -> ModGuts -> InScopeEnv -> Type -> IO (Either SDoc CoreExpr)
-buildDictionary env dflags guts inScope goalTy =
+buildDictionary :: HscEnv -> DynFlags -> ModGuts -> InScopeEnv -> Type -> CoreExpr -> Type -> IO (Either SDoc CoreExpr)
+buildDictionary env dflags guts inScope evType@(TyConApp tyCon evTypes) ev goalTy | isTupleTyCon tyCon =
+  reallyBuildDictionary env dflags guts inScope evType evTypes ev goalTy
+-- only 1-tuples in Haskell  
+buildDictionary env dflags guts inScope evType ev goalTy | isEvVarType evType =
+  reallyBuildDictionary env dflags guts inScope evType [evType] ev goalTy
+buildDictionary _env _dflags _guts _inScope evT _ev _goalTy = pprPanic "evidence type mismatch" (ppr evT)
+                                                         
+reallyBuildDictionary :: HscEnv -> DynFlags -> ModGuts -> InScopeEnv -> Type -> [Type] -> CoreExpr -> Type -> IO (Either SDoc CoreExpr)
+reallyBuildDictionary env dflags guts inScope evType evTypes ev goalTy =
   pprTrace' "\nbuildDictionary" (ppr goalTy) $
-  pprTrace' "buildDictionary in-scope evidence" (ppr (WithType . Var <$> uniqSetToList scopedDicts)) $
-  reassemble <$> buildDictionary' env dflags guts scopedDicts binder
+  pprTrace' "buildDictionary in-scope evidence" (ppr ev) $
+  reassemble <$> buildDictionary' env dflags guts evIdSet binder
  where
+   evIds = filter keepVar (map (\ (evTy, index) -> localId inScope ("evidence" ++ show index) evTy)  (zip evTypes [(0::Int)..]))
```

Comment:
Could you take a look at this alternative version, which tried to implement your suggestion from the comment with your name:

https://github.com/mikesperber/concat/commit/bc2911ebec73af77c0503dfe0da038f8fef7f938#diff-f51ebdf87e9d4ea860485c79e31c3ab3cb2ddce7cd2215f2c207a2901b01922cR137

This version delegates generating the identifier to `TcMType.newWanted`. You said that the comment is effectively overkill, but maybe it would solve this problem?

Reply:
Yes, I think that would avoid the problem I'm worried about here.

Reply:
@goldfirere So I did that, but the use you quote above remains. I can't see how that would be problematic though as the variables are bound and used within the expression returned by `buildDictionary` - and it does generate unique variables by including the index. Or am I overlooking something?

I also couldn't figure out how to implement the `UniqueSupply` suggestion - it seems the plugin would need to generate a new `UniqueSupply` for each invocation of the rule, but there's no way to communicate between the rule invocations. (I guess I could use `mkSplitUniqSupply`, but that seems really ugly.) Maybe I just don't understand the intended pragmatics of `UniqSupply` though?

Reply:
Is there a guarantee that each local id will have a unique name? It really depends on the semantics of `hashString` and `mkUniqueGrimily` (which really should not be used). The contract for `hashString` is (generally) that `s1 == s2` implies `hashString s1 == hashString s2`. But that's not what we need here; instead, we need `s1 /= s2` implies `hashString s1 /= hashString s2`, which cannot generally hold (given that `hashString` outputs an `Int32`). Given that you have precisely one call site of `localId` and that this call site does vary the argument string, you're probably safe in practice, but I wouldn't want this solution deployed anywhere, say, critical to life.

A quick fix (not using `UniqSupply`) would be to generate `evId`s one at a time, adding them to the in-scope set as you go. Then, if there is a hash collision, you'll detect it and rename away from it.

You should not generate a new `UniqSupply` at all. You can get a pre-made one from the `CoreM` monad, accessible from within `addRule`. Use `getUniqueSupplyM`. It should be OK to use the same `UniqSupply` for each invocation of the rule. Or maybe you tried this and ran into difficulty? Where?

Reply:
Thanks! Did this:

https://github.com/conal/concat/pull/65/commits/8fcf79d5006217b6f97b1ea92870ea4ee41616aa

I'm unsure I did this right though - there seem a myriad ways to construct unique variables throughout GHC.